block/internal/syncing/syncer.go

```diff
@@ -262,107 +251,118 @@ func (s *Syncer) processLoop() {
-	s.wg.Add(1)
-	go s.consumeNotifierEvents(s.dataSub, syncnotifier.EventTypeData)
-
-	return nil
-}
+const (
+	fastDAPollInterval  = 10 * time.Millisecond
+	futureHeightBackoff = 6 * time.Second // current celestia block time
+)
 
-func (s *Syncer) consumeNotifierEvents(sub *syncnotifier.Subscription, expected syncnotifier.EventType) {
+func (s *Syncer) daWorkerLoop() {
 	defer s.wg.Done()
 
 	if !s.waitForGenesis() {
 		return
 	}
 
-	logger := s.logger.With().Str("event_type", string(expected)).Logger()
-	logger.Info().Msg("starting P2P notifier worker")
-	defer logger.Info().Msg("P2P notifier worker stopped")
+	s.logger.Info().Msg("starting DA worker")
+	defer s.logger.Info().Msg("DA worker stopped")
+
+	nextDARequestAt := &time.Time{}
+	pollInterval := fastDAPollInterval
 
 	for {
+		s.tryFetchFromDA(nextDARequestAt)
 		select {
 		case <-s.ctx.Done():
 			return
-		case evt, ok := <-sub.C:
-			if !ok {
-				return
-			}
-			if evt.Type != expected {
-				logger.Debug().Str("received_type", string(evt.Type)).Msg("ignoring notifier event with unexpected type")
-				continue
-			}
-			s.logger.Debug().Str("event_type", string(evt.Type)).Str("source", string(evt.Source)).Uint64("height", evt.Height).Msg("received store event")
-			s.tryFetchFromP2P()
+		case <-time.After(pollInterval):
 		}
 	}
 }
 
-func (s *Syncer) daWorkerLoop() {
+func (s *Syncer) pendingWorkerLoop() {
 	defer s.wg.Done()
 
 	if !s.waitForGenesis() {
 		return
 	}
 
-	s.logger.Info().Msg("starting DA worker")
-	defer s.logger.Info().Msg("DA worker stopped")
+	s.logger.Info().Msg("starting pending worker")
+	defer s.logger.Info().Msg("pending worker stopped")
 
-	nextDARequestAt := &time.Time{}
-	pollInterval := min(10*time.Millisecond, s.config.Node.BlockTime.Duration)
-	if pollInterval <= 0 {
-		pollInterval = 10 * time.Millisecond
-	}
+	ticker := time.NewTicker(10 * time.Millisecond)
+	defer ticker.Stop()
 
 	for {
-		s.tryFetchFromDA(nextDARequestAt)
 		select {
 		case <-s.ctx.Done():
 			return
-		case <-time.After(pollInterval):
+		case <-ticker.C:
+			s.processPendingEvents()
 		}
 	}
 }
 
-func (s *Syncer) pendingWorkerLoop() {
+func (s *Syncer) p2pWorkerLoop() {
 	defer s.wg.Done()
 
 	if !s.waitForGenesis() {
 		return
 	}
 
-	s.logger.Info().Msg("starting pending worker")
-	defer s.logger.Info().Msg("pending worker stopped")
-
-	ticker := time.NewTicker(10 * time.Millisecond)
-	defer ticker.Stop()
+	logger := s.logger.With().Str("worker", "p2p").Logger()
+	logger.Info().Msg("starting P2P worker")
+	defer logger.Info().Msg("P2P worker stopped")
 
 	for {
 		select {
 		case <-s.ctx.Done():
 			return
-		case <-ticker.C:
-			s.processPendingEvents()
+		default:
+		}
+
+		currentHeight, err := s.store.Height(s.ctx)
+		if err != nil {
+			logger.Error().Err(err).Msg("failed to get current height for P2P worker")
+			if !s.sleepOrDone(50 * time.Millisecond) {
+				return
+			}
+			continue
+		}
+
+		targetHeight := currentHeight + 1
```

Comment:
The current height is loaded from store. When we have a slow consumer, the value would not be updated as fast as the tick triggers the p2pWorkerLoop. This can lead to duplicate pulls.
The cache deduplicates the events but may also skip the call as an optimization by tracking the last successful height fetched from p2p (store)